### ISSUE TITLE
Incorporating KL loss in the final loss

### ISSUE THREAD

**Issue by thechargedneutron** [XueZeyue/DanceGRPO#33]
Hi, thanks for the good work and the reproducible code. In the Hunyuan video training, it seems you are not using the KL loss for now. Do you think the model will learn better and be closer to the original model if you train with a KL loss?

I am looking to integrate KL loss in the current codebase. In Flow-GRPO paper, it is done here: https://github.com/yifan123/flow_grpo/blob/main/scripts/train_sd3.py#L922-L926. However, in your implementation, you are finetuning the whole transformer (is that correct?) and hence, you do not have a copy of the original transformer. My understanding is that we need to make a copy of the original transformer and then compute KL loss using that. Is that correct? Or are you storing the original transformer somewhere that can be used directly?

**Comment by XueZeyue** [comment]
Hi, I don't use KL loss for the following reasons:
1. In LLM, we use KL loss to avoid reward collapse because they are AR-style models. But for diffusion models, we don't observe reward collapse if we disable this loss.
2. For diffusion/flow-matching models, we should learn how to avoid reward over-optimization, which will reduce visual quality. For this, we have many engineering tricks, such as a) decreasing learning rate, b) enabling ema, c) KL loss, d) using multi-reward training. They are engineering tricks rather than research contributions. For VRAM efficiency, you can choose one of these tricks.
